File: app/src/main/java/com/example/mrakopediareader/pageslist/PageResultsAdapter.java
package com.example.mrakopediareader.pageslist;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mrakopediareader.R;
import com.example.mrakopediareader.api.Page;

import java.util.ArrayList;
import java.util.function.Consumer;

class PageResultsAdapter extends RecyclerView.Adapter<PageResultsAdapter.ViewHolder> {
    private ArrayList<Page> mDataset;

    private Consumer<Page> onClick;

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private LinearLayout layout;

        private Page page;

        public ViewHolder(LinearLayout layout, Consumer<Page> onClick) {
            super(layout);
            this.layout = layout;
            this.layout.setOnClickListener((view) -> {
                onClick.accept(this.page);
            });
        }

        public void setPage(Page page) {
            this.page = page;
            final TextView textView = this.layout.findViewById(R.id.pageTitle);
            textView.setText(this.page.getTitle());
        }
    }

    public PageResultsAdapter(ArrayList<Page> myDataset, Consumer<Page> onClick) {
        mDataset = myDataset;
        this.onClick = onClick;
    }

    @Override
    public PageResultsAdapter.ViewHolder onCreateViewHolder(
            @NonNull ViewGroup parent,
            int viewType
    ) {
        LinearLayout layout = (LinearLayout) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.search_result_view, parent, false);
        return new ViewHolder(layout, this.onClick);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final Page page = mDataset.get(position);
        holder.setPage(page);
    }

    @Override
    public int getItemCount() {
        return mDataset.size();
    }
}

File: app/src/main/java/com/example/mrakopediareader/pageslist/PageResultsAdapter.kt
package com.example.mrakopediareader.pageslist

import android.content.res.Resources
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.recyclerview.widget.RecyclerView
import com.example.mrakopediareader.R
import com.example.mrakopediareader.api.dto.Page
import com.example.mrakopediareader.api.dto.PageMetaInfo

internal class PageResultsAdapter(
    private val pages: List<Page>,
    private val resources: Resources,
    private val getPageMetaInfo: (title: String) -> PageMetaInfo?,
    private val onClick: (page: Page) -> Unit,
) : RecyclerView.Adapter<PagesListViewHolder>() {
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): PagesListViewHolder {
        val layout: LinearLayout? = LayoutInflater.from(parent.context)
            .inflate(R.layout.search_result_view, parent, false) as? LinearLayout
        return layout?.let {
            PagesListViewHolder(
                layout = it,
                getPageMetaInfo = getPageMetaInfo,
                resources = resources
            ) { page -> onClick(page) }
        } ?: throw Error("No Linear Layout")
    }

    override fun onBindViewHolder(holder: PagesListViewHolder, position: Int) {
        val page = pages[position]
        holder.setPage(page)
    }

    override fun getItemCount(): Int {
        return pages.size
    }
}